Lithium: a Rust-native back end

JakuJ opened the pull request:
This is a draft PR outlining the work on a new Rust-native back end for Prusti, developed in connection to my Master's thesis. The aim is to, among others, evaluate the performance benefits of going straight from CFG to SMT-LIB and decoupling from Viper.

The project is in its initial stages and so the code is highly volatile. Everything is subject to change. The purpose of this PR is to give Prusti developers an overview of the work being done in order to avoid double work, conflicts etc.

vakaras commented on prusti-server/src/process_verification.rs:
```diff
@@ -14,9 +14,88 @@ use prusti_common::{
+    // Create a new verifier each time.
+    // A workaround for https://github.com/viperproject/prusti-dev/issues/744
+    // TODO: Is it still needed?
+    let mut stopwatch = Stopwatch::start("prusti-server", "verifier startup");
```
`prusti-server` is used for accommodating the JVM startup time and should not be needed with the native backend.

JakuJ replied:
Yes, I intend to purge all my changes from `prusti-server` eventually, but for now I just plugged this in to make the compiler happy.

fpoli replied:
> `prusti-server` is used for accommodating the JVM startup time and should not be needed with the native backend.

Do you mean that the message is misleading? The server does provide caching and a uniform API to verify programs.

vakaras replied:
`prusti-server` is JVM dependent. If we want to use it for caching/network-based verification, I think we should create a server trait that is implemented by Viper and native servers.

fpoli replied:
Good point, the JVM dependency is bad.

vakaras commented on prusti-server/src/client.rs:
```diff
@@ -30,6 +30,27 @@ impl PrustiClient {
     pub async fn verify(
         &self,
         request: VerificationRequest,
+    ) -> reqwest::Result<VerificationResult> {
```
I suggest creating a version of `VerificationResult` that is not Viper dependent and using it here instead.

JakuJ replied:
I moved the `VerificationResult` struct to another crate `backend-common`. The `viper` crate and my (future) native verifier crate depend on it.

vakaras commented on prusti-server/src/client.rs:
```diff
@@ -30,6 +30,27 @@ impl PrustiClient {
         request: VerificationRequest,
+    ) -> reqwest::Result<VerificationResult> {
+        let response = if let VerificationBackend::Lithium = request.backend_config.backend {
+            self.verify_natively(request).await
```
Instead of branching on the client side, I would suggest branching on the server side because this allows doing (at least theoretically) the heavy verification task on a different machine. (The user can still disable the Prusti server completely if they want to verify locally and avoid the serialization overhead.)

vakaras commented on prusti-server/src/process_verification.rs:
```diff
@@ -104,11 +183,21 @@ pub fn process_verification_request<'v, 't: 'v>(
+        let mut result = if let VerificationBackend::Lithium = backend {
+            if let prusti_common::vir::program::Program::Low(ref low_program) = request.program {
+                stopwatch.start_next("verification");
+                verifier.verify_vir(low_program)
```
I would suggest having a `Verifier` trait that is implemented by `ViperVerifier` and `NativeVerifier`. That trait then would have a `verify` method that accepts a VIR program, which `ViperVerifier` would internally convert to a Viper program.

JakuJ replied:
All test projects expect a `verify` method that accepts a Viper program. I exposed such method on the `JavaVerifier` so that nothing needs to be renamed in test projects. The method I placed on the trait is named `verify_vir` and accepts a `Program::Legacy` or `Program::Low`, with the `JavaVerifier` transforming it to Viper before passing it to its `verify` method.

vakaras replied:
Since the test projects are expected to construct the `JavaVerifier` directly, this sounds good to me.

vakaras commented on viper/src/verifier.rs:
```diff
@@ -4,345 +4,14 @@
-        })
-    }
-}
+    fn verify(&mut self, program: crate::Program) -> VerificationResult;
```
I would remove this method because this way, you could have the native backend completely independent of Viper.

JakuJ committed:
Add Lithium back end

JakuJ committed:
Add Lithium to VerificationBackend enum

JakuJ committed:
Fix simple branching (if-statements in low vir)

JakuJ committed:
Wonky way of reporting offending positions

JakuJ committed:
Fix branching assumptions for switchInt

JakuJ committed:
Fix "otherwise" branch of switchInt statements

JakuJ committed:
Fix after rebase

JakuJ committed:
Add tests for loops, remove empty Conditional statements

JakuJ committed:
Fix mod operator

JakuJ committed:
Add initial F32 and F64 support

JakuJ committed:
Use FP operators, choose correct type declaration

JakuJ committed:
Tests for floating-point support

JakuJ committed:
Add preamble for CVC5, split fp.neq into 32 and 64-bit variants

JakuJ committed:
Fixes after merge

JakuJ committed:
Quantifiers to snapshots, fix unsized type bound checking bug

JakuJ committed:
Hack quantifier generation for complex types

JakuJ committed:
Fix after rebase

JakuJ committed:
Fix lowering for quantifiers that use closures over function arguments

JakuJ committed:
Add sum-of-primes test file

JakuJ committed:
Add super sale test case

JakuJ committed:
Update test cases

JakuJ committed:
Add Euclid test case, fix quantifiers type mismatch

JakuJ committed:
Optimization module

JakuJ committed:
Correct reporting of failure reason positions

JakuJ committed:
Remove redundant reasons for failure if the same as failing place

JakuJ committed:
FP != for Viper

JakuJ committed:
Passing FP test case

JakuJ committed:
Remove main.rs from Lithium

JakuJ committed:
Commit the config

JakuJ committed:
Use files, not stdio, fix negative int encoding

JakuJ committed:
Make clippy happy

JakuJ committed:
cargo fmt

JakuJ committed:
More cargo fmt

JakuJ committed:
Increase SMT timeout, comment out problematic code in error_manager

JakuJ committed:
Fix abs(i32::MIN) overflowing

JakuJ committed:
Refactor encoding of Statement::Conditional

JakuJ committed:
Fix invalid file paths when emitting SMT files

JakuJ committed:
Fix quantifier optimization rules

JakuJ committed:
Various bugfixes

JakuJ committed:
Make sure positions are preserved by the optimizer

JakuJ committed:
Add bisect test case

JakuJ committed:
Remove overcomplicated tests

JakuJ committed:
Fix test error placement

JakuJ committed:
Clean up SMT preambles, fix modulo operator encoding

JakuJ committed:
Add Lithium test coverage script

To be removed when we want to merge this into Prusti. Committing for now since it's relevant to my thesis.

JakuJ committed:
Fixes after merging master

JakuJ committed:
Make clippy happy

JakuJ committed:
Fix test compilation error

JakuJ committed:
Restore config

JakuJ committed:
Fixes after rebase

JakuJ committed:
Clippy

fpoli commented on prusti-server/src/backend.rs:
```diff
@@ -36,6 +38,10 @@ impl<'a> Backend<'a> {
                 })
             }
+            Backend::Lithium(lithium) => {
+                Stopwatch::start("prusti-server", "vir verification");
```
Without this, the stopwatch is immediately dropped and does *not* measure how long the verification takes.
```suggestion
                let _stopwatch = Stopwatch::start("prusti-server", "vir verification");
```

JakuJ committed:
Fix stopwatch getting released immediately

JakuJ committed:
Merge branch 'master' into lithium

JakuJ committed:
Add PRUSTI_NATIVE_VERIFIER env variable

JakuJ committed:
Restore test case unrelated to Lithium

fpoli commented on test.sh:
Could you move this script into the `scripts` folder?

fpoli commented on prusti-utils/src/config.rs:
```diff
@@ -190,6 +191,17 @@ lazy_static::lazy_static! {
             allowed_keys.iter().filter(|key| key.len() > MAX_CONFIG_LEN).collect::<Vec<_>>()
         );
 
+        // 0. Apply Lithium overrides if necessary
+        if let Ok(true) = env::var("PRUSTI_NATIVE_VERIFIER").map(|s| s.parse::<bool>().unwrap_or(false)) {
+            settings.set_default("viper_backend", "Lithium").unwrap();
+            settings.set_default("verify_specifications_backend", "Lithium").unwrap();
+            settings.set_default("encode_bitvectors", true).unwrap();
+            settings.set_default("enable_purification_optimization", true).unwrap();
+            settings.set_default("unsafe_core_proof", true).unwrap();
+            settings.set_default("verify_core_proof", false).unwrap();
+            settings.set_default("inline_caller_for", true).unwrap();
+        }
+
```
What is the use case of this new `PRUSTI_NATIVE_VERIFIER`?

JakuJ replied:
Grouping all required flags into one, to facilitate switching between the backends. Alternatively I could check if `viper_backend == "Lithium"`.

Why an environment variable? For some reason if I do

```rust
if settings.get_bool("native_verifier").unwrap_or(false) {
   ...
}
```

at the end (after the `-P` arguments) and provide it via `cargo-prusti -- -Pnative_verifier=true` it breaks everything, including the building of `prusti-contracts` 🤨

fpoli replied:
There are various reasons why I'd suggest to undo this change and look for another solution:
* The `native_verifier` and `viper_backend` flags are partially overlapping.
* The code checks the environment variable `PRUSTI_NATIVE_VERIFIER`, but does not work if the same flag is set via `-P`. This goes against the current design of the configuration flags.
* Changing the default of some flags depending on the value of some flags looks like the kind of thing that easily leads to surprises and hard-to-debug issues. We were forced to do it in one case (`DEFAULT_PRUSTI_*`) because of the way cargo works, but in this PR I think that we have more freedom to choose a better solution.

If Lithium *requires* a specific set of flags, what about just implementing runtime checks like we do [here](https://github.com/viperproject/prusti-dev/blob/a5c29c994cee03e1ba02c3bc2c2761803571d3f5/prusti-utils/src/config.rs#L602-L605)? Moreover, if a set of tests require a long combination of flags, what about adding one entry [here](https://github.com/viperproject/prusti-dev/blob/a5c29c994cee03e1ba02c3bc2c2761803571d3f5/prusti-tests/tests/compiletest.rs#L162-L219), with a new test folder such as `prusti-tests/tests/verify_lithium`?
